### PR TITLE
3 new functions added to `y_weapondata` 

### PR DESCRIPTION
**Pull request by DEntis-T**
Reopen the request #609.

### CODE REVIEW

**Comment by DEntis-T** [comment]
@Y-Less

File: YSI_Game/y_weapondata/y_weapondata_entry.inc [hunk 1]
```diff
@@ -87,32 +82,42 @@ Optional plugins:
...
+
+stock YSI_Weapon_IsValid__(weaponid)
 {
-	return (weaponid >= 1 && weaponid <= 18 || weaponid >= 21 && weaponid <= 46);
+	switch(weaponid)
+	{
+		case 1 .. 18:
+			return true;
+		case 21 .. 46:
+			return true;
```
Comment (Y-Less) [inline YSI_Game/y_weapondata/y_weapondata_entry.inc]: The `if` is better, this is even better:

```pawn
return (1<= weaponid <= 18) || (21<= weaponid >= 21 <= 46);
```

File: YSI_Game/y_weapondata/y_weapondata_entry.inc [hunk 2]
```diff
@@ -65,20 +63,17 @@ Thanks:
 		me to strive to better.
 	Pixels^ - Running XScripters where the idea was born.
 	Matite - Pestering me to release it and using it.
-
 Very special thanks to:
 	Thiadmer - PAWN, whose limits continue to amaze me!
 	Kye/Kalcor - SA:MP.
 	SA:MP Team past, present and future - SA:MP.
-
```
Comment (Y-Less) [inline YSI_Game/y_weapondata/y_weapondata_entry.inc]: Lines are missing.

File: YSI_Game/y_weapondata/y_weapondata_entry.inc [hunk 3]
```diff
@@ -87,32 +82,42 @@ Optional plugins:
...
+#define WEAPONDATA_300MAGNUM			".300 Magnum"
+#define WEAPONDATA_ROCKET				"Rocket"
+#define WEAPONDATA_INVALID_RETURN 		-1
+
+#define WEAPON_FIRE_TYPE_MELEE 			0
+#define WEAPON_FIRE_TYPE_PROJECTILE 	1
+#define WEAPON_FIRE_TYPE_INSTANT_HIT 	2
+#define WEAPON_FIRE_TYPE_AREA_EFFECT 	3
+#define WEAPON_FIRE_TYPE_USE 			4
+#define WEAPON_FIRE_TYPE_CAMERA 		5
```
Comment (Y-Less) [inline YSI_Game/y_weapondata/y_weapondata_entry.inc]: You still haven't explained types.

File: YSI_Game/y_weapondata/y_weapondata_entry.inc [hunk 4]
```diff
@@ -87,32 +82,42 @@ Optional plugins:
...
+			return true;
+		case 21 .. 46:
+			return true;
+	}
+	return false;
 }
 
+// Many libraries already have `Weapon_IsValid` defined.
+#if !defined Weapon_IsValid
+	#define Weapon_IsValid(%0) (%0>=1&&%0<=18||%0>=21&&%0<=46)
```
Comment (Y-Less) [inline YSI_Game/y_weapondata/y_weapondata_entry.inc]: Don't use a macro for this, they are broken.

File: YSI_Game/y_weapondata/y_weapondata_entry.inc [hunk 5]
```diff
@@ -204,7 +209,7 @@ static stock YSI_g_WeaponFireType[47] = {
 
 stock Weapon_GetFireType(weaponid)
 {
-	return Weapon_IsValid(weaponid) ? YSI_g_WeaponFireType[weaponid] : (-1);
+	return YSI_Weapon_IsValid__(weaponid) ? YSI_g_WeaponFireType[weaponid] : (-1);
```
Comment (Y-Less) [inline YSI_Game/y_weapondata/y_weapondata_entry.inc]: Just use `Weapon_IsValid`

**Review by Y-Less — CHANGES_REQUESTED** [review]
I've asked you about five times why "types" are different to "slots".  They map 1-to-1, so there's absolutely no need for two things giving the same information.  Why do they exist?  What do they add?  Where does the idea come from?  Why does just using slots not work?

File: YSI_Game/y_weapondata/y_weapondata_entry.inc (hunk 3 above)
Comment (DEntis-T) [reply YSI_Game/y_weapondata/y_weapondata_entry.inc]: Those are fire types, how the certain weapon does damage or effect

**Comment by DEntis-T** [comment]
> I've asked you about five times why "types" are different to "slots". They map 1-to-1, so there's absolutely no need for two things giving the same information. Why do they exist? What do they add? Where does the idea come from? Why does just using slots not work?

They don't map 1-to-1; look:
```pawn
stock WeaponType:Weapon_GetType(weaponid)
{ 
	return WeaponType:(Weapon_GetSlot(weaponid) - 1);
}
```
There's not a slot 0.

File: YSI_Game/y_weapondata/y_weapondata_entry.inc (hunk 5 above)
Comment (DEntis-T) [reply YSI_Game/y_weapondata/y_weapondata_entry.inc]: For some reason `stock Weapon_IsValid` gives an invalid instruction run-time error. It only works if it is a macro.

File: YSI_Game/y_weapondata/y_weapondata_entry.inc (hunk 4 above)
Comment (DEntis-T) [reply YSI_Game/y_weapondata/y_weapondata_entry.inc]: For some reason `stock Weapon_IsValid` gives an invalid instruction run-time error. It only works if it is a macro.

File: YSI_Game/y_weapondata/y_weapondata_entry.inc (hunk 2 above)
Comment (DEntis-T) [reply YSI_Game/y_weapondata/y_weapondata_entry.inc]: That's peculiar.

File: YSI_Game/y_weapondata/y_weapondata_entry.inc (hunk 1 above)
Comment (DEntis-T) [reply YSI_Game/y_weapondata/y_weapondata_entry.inc]: For some reason `stock Weapon_IsValid` gives an invalid instruction run-time error. It only works if it is a macro.

**Pull request closed without merging**